src/com/company/Shop.java

```diff
@@ -53,19 +86,74 @@ public void fire(Employee employee) {
         employees.remove(employee);
     }
 
-    public void addProduct(ProductDetails product, double quantity) {
+
+    public void addProduct(Product product, double quantity) {
+        boolean isFound = false;
         for (int i = 0; i < products.size(); i++) {
-            ProductDetails productDetails = products.get(i);
-            if (productDetails.equals(product)) {
-                product.quantity += quantity;
-            } else {
-                products.add(product);
+            ProductDetails currentProduct = products.get(i);
+            if (currentProduct.getProduct().getId() == product.getId()) {
+                currentProduct.quantity += quantity;
+                isFound = true;
             }
         }
+        if(!isFound){
+            products.add(new ProductDetails(product , quantity));
+        }
+    }
+
+    public void deleteProduct(int productId) {
+        ProductDetails toRemove = null;
+        for (ProductDetails product : products) {
+            if (product.getProduct().getId() == productId) {
+                toRemove = product;
+            }
+        }
+        if(toRemove!=null){
+            products.remove(toRemove);
+        }
+    }
+    public ProductDetails findProductById(int productId){
+
+        for(ProductDetails productDetails : products){
+            if(productDetails.getProduct().getId() == productId) {
+                return productDetails;
+            }
+        }
+        return null;
+
+    }
+
+    public void updateProduct(int productId, Product modifiedProduct) {
+        ProductDetails productDetails = findProductById(productId);
+        deleteProduct(productId);
+        modifiedProduct.setId(productId);
+        addProduct(modifiedProduct , productDetails.quantity);
     }
 
     public void addService(Service service) {
         services.add(service);
     }
 
+    public void deleteService(int serviceId) {
+
+        for (Service service : services) {
```

Comment:
we shouldn't modify (add/remove) a list during iteration. you can use `findServiceById` to first find a service and then remove